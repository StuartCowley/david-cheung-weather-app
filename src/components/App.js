import React, { useState, useEffect } from "react";
import LocationDetails from "./LocationDetails";
import ForecastSummaries from "./ForecastSummaries";
import ForecastDetails from "./ForecastDetails";
import SearchForm from "./SearchForm";
import DigitalClock from "./DigitalClock";
import getForecast from "../requests/getForecast";
import "../styles/app.css";

function App() {
  const [forecasts, setForecasts] = useState([]);
  const [location, setLocation] = useState({
    city: "",
    country: "",
  });
  const [selectedDate, setSelectedDate] = useState(0);
  const [searchText, setSearchText] = useState("");
  const [errorMessage, setErrorMessage] = useState("");
  const [dayMode, setDayMode] = useState("day-mode");

  const selectedForecast = forecasts.find(
    (forecast) => forecast.date === selectedDate
  );

  const handleForecastSelect = (date) => {
    setSelectedDate(date);
  };

  useEffect(() => {
    getForecast(
      searchText,
      setSelectedDate,
      setForecasts,
      setLocation,
      setErrorMessage
    );
  }, []); // eslint-disable-line react-hooks/exhaustive-deps

  const handleCitySearch = () => {
    getForecast(
      searchText,
      setSelectedDate,
      setForecasts,
      setLocation,
      setErrorMessage
    );
  };

  return (
    <div className={`weather-app ${dayMode}`}>
      <h1>FIVE DAY FORECAST</h1>

      <button
        className="day-night-mode"
        type="button"
        onClick={() => {
          setDayMode(dayMode === "day-mode" ? "night-mode" : "day-mode");
        }}
      >
        D/N
      </button>

      <LocationDetails
        city={location.city}
        country={location.country}
        errorMessage={errorMessage}
      />

      <SearchForm
        searchText={searchText}
        setSearchText={setSearchText}
        onSubmit={handleCitySearch}
      />

      {forecasts.length > 0 && (
        <>
          <ForecastSummaries
            forecasts={forecasts}
            onForecastSelect={handleForecastSelect}
          />
          {selectedForecast && <ForecastDetails forecast={selectedForecast} />}
        </>
      )}
      <DigitalClock />
    </div>
  );
}

export default App;
