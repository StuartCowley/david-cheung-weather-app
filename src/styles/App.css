.weather-app {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 20px;
  min-height: 600px;
  background-repeat: no-repeat;
  background-size: cover;
}

.day-mode {
  color: black;
  background-image: url("https://s.facegfx.com/image/2014/7/5/tree-with-four-seasons-vector-04-free.jpg");
}

.night-mode {
  color: white;
  background-image: linear-gradient(rgba(0, 0, 0, 0.527), rgba(0, 0, 0, 0.5)),
    url("https://s.facegfx.com/image/2014/7/5/tree-with-four-seasons-vector-04-free.jpg");
}

.day-night-mode {
  position: absolute;
  top: 10px;
  right: 10px;
  border-radius: 50%;
  background-color: rgba(0, 0, 0, 0);
}

.day-night-mode:hover {
  background-color: greenyellow;
}

h1 {
  margin: 5px;
}

h3 {
  padding: 5px;
  width: 100%;
  text-align: center;
}

.forecast-summaries {
  display: flex;
  width: 100%;
  padding: 20px;
}
