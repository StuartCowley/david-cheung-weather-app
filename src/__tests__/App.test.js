import { render, screen, waitFor } from "@testing-library/react";
import App from "../components/App";
import renderer from "react-test-renderer";
import axios from "axios";

describe("App", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("Renders as expected", () => {
    Date.now = jest.fn(() => 1482363367071);
    const rendered = renderer.create(<App />);

    expect(rendered).toMatchSnapshot();
  });

  test("Assert h1 is present", () => {
    render(<App />);

    const headings = screen.getAllByRole("heading");
    expect(headings[0]).toHaveTextContent("FIVE DAY FORECAST");
  });

  test("renders learn react link with spy axios", async () => {
    const mAxiosResponse = {
      data: {
        location: {
          city: "testCity",
          country: "testCountry",
        },
        forecasts: [
          {
            date: 1525046400000,
            temperature: {
              max: 11,
              min: 4,
            },
            wind: {
              speed: 13,
              direction: "s",
            },
            humidity: 30,
            description: "Clear",
            icon: 800,
          },
          {
            date: 1525132800000,
            temperature: {
              max: 13,
              min: 8,
            },
            wind: {
              speed: 60,
              direction: "ne",
            },
            humidity: 80,
            description: "Stormy",
            icon: 211,
          },
          {
            date: 1525219200000,
            temperature: {
              max: 1,
              min: -2,
            },
            wind: {
              speed: 5,
              direction: "n",
            },
            humidity: 50,
            description: "Heavy Snow",
            icon: 602,
          },
          {
            date: 1525305600000,
            temperature: {
              max: 20,
              min: 4,
            },
            wind: {
              speed: 150,
              direction: "e",
            },
            humidity: 80,
            description: "Tornado",
            icon: 781,
          },
          {
            date: 1525392000000,
            temperature: {
              max: 25,
              min: 18,
            },
            wind: {
              speed: 8,
              direction: "nne",
            },
            humidity: 50,
            description: "Hazy",
            icon: 721,
          },
        ],
      },
    };
    jest.spyOn(axios, "get").mockResolvedValue(mAxiosResponse);
    render(<App />);
    await waitFor(() => {
      expect(screen.getByText("testCity, testCountry")).toBeTruthy();
      expect(screen.getByText("testCity, testCountry")).toBeInstanceOf(
        HTMLHeadingElement
      );
    });
  });
});
