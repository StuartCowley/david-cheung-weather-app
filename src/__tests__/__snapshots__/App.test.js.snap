// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Renders as expected 1`] = `
<div
  className="weather-app day-mode"
>
  <h1>
    FIVE DAY FORECAST
  </h1>
  <button
    className="day-night-mode"
    onClick={[Function]}
    type="button"
  >
    D/N
  </button>
  <h1
    className="location-details"
  >
    , 
  </h1>
  <div
    className="search-form"
  >
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      onClick={[Function]}
      type="submit"
    >
      Search
    </button>
  </div>
  <div
    className="digital-clock"
  >
     
    11:36:07 PM
     
  </div>
</div>
`;
